File: javascript/packages/linter/src/rules/erb-no-unsafe-script-interpolation.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { PrismVisitor } from "@herb-tools/core"

import {
  getTagLocalName,
  getAttribute,
  getStaticAttributeValue,
  isERBNode,
  isERBOutputNode,
  isHTMLOpenTagNode,
} from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, HTMLElementNode, Node, ERBContentNode, ParserOptions, PrismNode } from "@herb-tools/core"

const SAFE_METHOD_NAMES = new Set([
  "to_json",
  "json_escape",
])

const ESCAPE_JAVASCRIPT_METHOD_NAMES = new Set([
  "j",
  "escape_javascript",
])

class SafeCallDetector extends PrismVisitor {
  public hasSafeCall = false
  public hasEscapeJavascriptCall = false

  visitCallNode(node: PrismNode): void {
    if (SAFE_METHOD_NAMES.has(node.name)) {
      this.hasSafeCall = true
    }

    if (ESCAPE_JAVASCRIPT_METHOD_NAMES.has(node.name)) {
      this.hasEscapeJavascriptCall = true
    }

    this.visitChildNodes(node)
  }
}

class ERBNoUnsafeScriptInterpolationVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    if (!isHTMLOpenTagNode(node.open_tag)) {
      super.visitHTMLElementNode(node)
      return
    }

    if (getTagLocalName(node.open_tag) === "script") {
      this.checkScriptElement(node)
    }

    super.visitHTMLElementNode(node)
  }

  private checkScriptElement(node: HTMLElementNode): void {
    if (!isHTMLOpenTagNode(node.open_tag)) return

    const typeAttribute = getAttribute(node.open_tag, "type")
    const typeValue = typeAttribute ? getStaticAttributeValue(typeAttribute) : null

    if (typeValue === "text/html") return
    if (!node.body || node.body.length === 0) return

    this.checkNodesForUnsafeOutput(node.body)
  }

  private checkNodesForUnsafeOutput(nodes: Node[]): void {
    for (const child of nodes) {
      if (!isERBNode(child)) continue
      if (!isERBOutputNode(child)) continue

      const erbContent = child as ERBContentNode
      const prismNode = erbContent.prismNode
      const detector = new SafeCallDetector()

      if (prismNode) detector.visit(prismNode)
      if (detector.hasSafeCall) continue

      if (detector.hasEscapeJavascriptCall) {
        this.addOffense(
          "Avoid `j()` / `escape_javascript()` in `<script>` tags. It is only safe inside quoted string literals. Use `.to_json` instead, which is safe in any position.",
          child.location,
        )

        continue
      }

      this.addOffense(
        "Unsafe ERB output in `<script>` tag. Use `.to_json` to safely serialize values into JavaScript.",
        child.location,
      )
    }
  }
}

export class ERBNoUnsafeScriptInterpolationRule extends ParserRule {
  static ruleName = "erb-no-unsafe-script-interpolation"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoUnsafeScriptInterpolationVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_unsafe_script_interpolation.rs
use crate::utils::erb_utils::is_output_tag_opening;
use crate::utils::prism_utils::walk_prism;
use crate::utils::tag_utils::{get_attribute, get_open_tag, get_static_attribute_value};

use herb::nodes::{AnyNode, HTMLElementNode};
use herb::prism::PrismNode;
use herb::Visitor;

const SAFE_METHOD_NAMES: &[&str] = &["to_json", "json_escape"];
const ESCAPE_JAVASCRIPT_METHOD_NAMES: &[&str] = &["j", "escape_javascript"];

rule_visitor!(ERBNoUnsafeScriptInterpolationVisitor);
define_parser_rule!(
  ERBNoUnsafeScriptInterpolationRule,
  "erb-no-unsafe-script-interpolation",
  Error,
  ERBNoUnsafeScriptInterpolationVisitor,
  parser_options: { prism_nodes: true },
  introduced_in: "0.9.0"
);

#[derive(Default)]
struct SafeCallDetector {
  has_safe_call: bool,
  has_escape_javascript_call: bool,
}

impl SafeCallDetector {
  fn detect(prism_node: &PrismNode) -> Self {
    let mut detector = Self::default();

    walk_prism(prism_node, &mut |node| {
      if node.is("CallNode") {
        if let Some(name) = node.name.as_deref() {
          if SAFE_METHOD_NAMES.contains(&name) {
            detector.has_safe_call = true;
          }

          if ESCAPE_JAVASCRIPT_METHOD_NAMES.contains(&name) {
            detector.has_escape_javascript_call = true;
          }
        }
      }

      true
    });

    detector
  }
}

impl ERBNoUnsafeScriptInterpolationVisitor {
  fn check_nodes_for_unsafe_output(&mut self, nodes: &[AnyNode]) {
    for child in nodes {
      let (tag_opening_token, location, prism_node) = match child {
        AnyNode::ERBContentNode(erb) => (erb.tag_opening.as_ref(), &erb.location, erb.prism()),
        AnyNode::ERBBlockNode(erb) => (erb.tag_opening.as_ref(), &erb.location, erb.prism()),
        AnyNode::ERBRenderNode(erb) => (erb.tag_opening.as_ref(), &erb.location, erb.prism()),
        AnyNode::ERBYieldNode(erb) => (erb.tag_opening.as_ref(), &erb.location, None),
        _ => continue,
      };

      let tag_opening = tag_opening_token.map(|token| token.value.as_str()).unwrap_or("");

      if !is_output_tag_opening(tag_opening) {
        continue;
      }

      let detector = match prism_node {
        Some(prism_node) => SafeCallDetector::detect(prism_node),
        None => SafeCallDetector::default(),
      };

      if detector.has_safe_call {
        continue;
      }

      if detector.has_escape_javascript_call {
        self.add_offense(
          "Avoid `j()` / `escape_javascript()` in `<script>` tags. It is only safe inside quoted string literals. Use `.to_json` instead, which is safe in any position.",
          location.clone(),
        );

        continue;
      }

      self.add_offense(
        "Unsafe ERB output in `<script>` tag. Use `.to_json` to safely serialize values into JavaScript.",
        location.clone(),
      );
    }
  }
}

impl Visitor for ERBNoUnsafeScriptInterpolationVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if let Some(open_tag) = get_open_tag(node) {
      let is_script = open_tag.tag_name.as_ref().map(|token| token.value.to_lowercase()).as_deref() == Some("script");

      let is_html_template = get_attribute(open_tag, "type")
        .and_then(get_static_attribute_value)
        .map(|value| value == "text/html")
        .unwrap_or(false);

      if is_script && !is_html_template {
        self.check_nodes_for_unsafe_output(&node.body);
      }
    }

    self.walk_html_element_node(node);
  }
}
